File: packages/agent/contracts.py
"""Versioned, closed contracts shared by agent components and audit storage."""

from __future__ import annotations

from datetime import UTC, datetime
from enum import StrEnum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field, model_validator

SCHEMA_VERSION = "1.0.0"


def utc_now() -> datetime:
    return datetime.now(UTC)


class ClosedModel(BaseModel):
    """Reject unknown fields so malformed planner output cannot reach an executor."""

    model_config = ConfigDict(extra="forbid", use_enum_values=False)


class InputModality(StrEnum):
    TEXT = "text"
    VOICE_TRANSCRIPT = "voice_transcript"


class ActionType(StrEnum):
    NAVIGATE = "navigate"
    CLICK = "click"
    TYPE = "type"
    SELECT = "select"
    CHECK = "check"
    UNCHECK = "uncheck"
    PRESS = "press"
    SCROLL = "scroll"
    WAIT = "wait"
    BACK = "back"
    ASK_USER = "ask_user"
    HANDOFF = "handoff"
    STOP = "stop"


class VerificationStatus(StrEnum):
    UNVERIFIED = "UNVERIFIED"
    VERIFIED = "VERIFIED"
    FAILED = "FAILED"
    INCONCLUSIVE = "INCONCLUSIVE"
    STALE = "STALE"


class RiskLevel(StrEnum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class TerminalReason(StrEnum):
    COMPLETED = "COMPLETED"
    USER_STOP = "USER_STOP"
    MAX_STEPS = "MAX_STEPS"
    SAFETY_STOP = "SAFETY_STOP"
    ERROR = "ERROR"


class ErrorCode(StrEnum):
    NONE = "NONE"
    INVALID_ACTION = "INVALID_ACTION"
    TARGET_NOT_FOUND = "TARGET_NOT_FOUND"
    STALE_OBSERVATION = "STALE_OBSERVATION"
    EXECUTION_FAILED = "EXECUTION_FAILED"
    VERIFICATION_FAILED = "VERIFICATION_FAILED"
    APPROVAL_REQUIRED = "APPROVAL_REQUIRED"
    USER_TAKEOVER = "USER_TAKEOVER"
    MAX_STEPS_REACHED = "MAX_STEPS_REACHED"
    INTERNAL_ERROR = "INTERNAL_ERROR"


class HandoffStatus(StrEnum):
    NONE = "NONE"
    REQUESTED = "REQUESTED"
    ACTIVE = "ACTIVE"
    RESUMING = "RESUMING"
    COMPLETED = "COMPLETED"


class MessageRole(StrEnum):
    USER = "USER"
    AGENT = "AGENT"
    SYSTEM = "SYSTEM"


class InterventionKind(StrEnum):
    APPROVAL = "APPROVAL"
    TAKEOVER = "TAKEOVER"
    CLARIFICATION = "CLARIFICATION"
    CANCEL = "CANCEL"


class InterventionStatus(StrEnum):
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    COMPLETED = "COMPLETED"


class UserCommand(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    command_id: UUID = Field(default_factory=uuid4)
    session_id: UUID
    text: str = Field(min_length=1, max_length=4_000)
    modality: InputModality = InputModality.TEXT
    received_at: datetime = Field(default_factory=utc_now)


class GoalSpec(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    task_id: str = Field(pattern=r"^T(?:0[1-9]|1[0-2])$")
    objective: str = Field(min_length=1, max_length=4_000)
    constraints: list[str] = Field(default_factory=list, max_length=32)
    allowed_actions: list[ActionType] = Field(min_length=1)
    max_steps: int = Field(ge=1, le=100)
    risk_level: RiskLevel


class AXNode(ClosedModel):
    node_id: str = Field(min_length=1, max_length=256)
    role: str = Field(min_length=1, max_length=128)
    name: str = Field(default="", max_length=1_000)
    description: str | None = Field(default=None, max_length=2_000)
    value_summary: str | None = Field(default=None, max_length=1_000)
    disabled: bool = False
    focused: bool = False
    children: list[str] = Field(default_factory=list)


class Observation(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    observation_ref: UUID = Field(default_factory=uuid4)
    version: int = Field(ge=1)
    url: str = Field(min_length=1, max_length=4_000)
    title: str = Field(default="", max_length=1_000)
    captured_at: datetime = Field(default_factory=utc_now)
    nodes: list[AXNode] = Field(default_factory=list)
    focused_node_id: str | None = None


class AgentAction(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    step_id: UUID = Field(default_factory=uuid4)
    action_type: ActionType
    target_ref: str | None = Field(default=None, max_length=256)
    input_value: str | None = Field(default=None, max_length=4_000)
    key: str | None = Field(default=None, max_length=64)
    observation_version: int = Field(ge=1)
    expected_effect: str = Field(min_length=1, max_length=2_000)
    risk_level: RiskLevel = RiskLevel.LOW
    requires_approval: bool = False

    @model_validator(mode="after")
    def validate_action_shape(self) -> AgentAction:
        target_actions = {
            ActionType.CLICK,
            ActionType.TYPE,
            ActionType.SELECT,
            ActionType.CHECK,
            ActionType.UNCHECK,
        }
        if self.action_type in target_actions and not self.target_ref:
            raise ValueError(f"target_ref wajib untuk action {self.action_type.value}")
        if self.action_type in {ActionType.TYPE, ActionType.SELECT} and self.input_value is None:
            raise ValueError(f"input_value wajib untuk action {self.action_type.value}")
        if self.action_type is ActionType.PRESS and not self.key:
            raise ValueError("key wajib untuk action press")
        if self.risk_level is RiskLevel.HIGH and not self.requires_approval:
            raise ValueError("action HIGH wajib requires_approval=true")
        return self


class VerificationResult(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    verification_id: UUID = Field(default_factory=uuid4)
    step_id: UUID
    status: VerificationStatus
    evidence: list[str] = Field(default_factory=list, max_length=32)
    before_observation_ref: UUID
    after_observation_ref: UUID | None = None
    checked_at: datetime = Field(default_factory=utc_now)
    error_code: ErrorCode = ErrorCode.NONE

    @model_validator(mode="after")
    def verified_requires_after_evidence(self) -> VerificationResult:
        if self.status is VerificationStatus.VERIFIED and (
            self.after_observation_ref is None or not self.evidence
        ):
            raise ValueError("VERIFIED wajib memiliki after_observation_ref dan evidence")
        return self


class RelevantItem(ClosedModel):
    semantic_ref: str = Field(min_length=1, max_length=256)
    label: str = Field(min_length=1, max_length=1_000)
    reason: str = Field(min_length=1, max_length=2_000)
    observation_version: int = Field(ge=1)


class TaskMapSnapshot(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    snapshot_id: UUID = Field(default_factory=uuid4)
    session_id: UUID
    run_id: UUID
    version: int = Field(ge=1)
    observation_version: int = Field(ge=1)
    completed_items: list[str] = Field(default_factory=list)
    pending_items: list[str] = Field(default_factory=list)
    relevant_items: list[RelevantItem] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=utc_now)


class FocusHandoff(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    handoff_id: UUID = Field(default_factory=uuid4)
    run_id: UUID
    status: HandoffStatus
    target_ref: str | None = Field(default=None, max_length=256)
    announcement: str = Field(min_length=1, max_length=2_000)
    created_at: datetime = Field(default_factory=utc_now)


class AgentState(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    session_id: UUID
    thread_id: str = Field(min_length=1, max_length=256)
    run_id: UUID
    task_id: str = Field(pattern=r"^T(?:0[1-9]|1[0-2])$")
    goal: GoalSpec
    constraints: list[str] = Field(default_factory=list)
    observation_version: int = Field(default=0, ge=0)
    task_map_version: int = Field(default=0, ge=0)
    active_semantic_ref: str | None = None
    verification: VerificationResult | None = None
    handoff_status: HandoffStatus = HandoffStatus.NONE
    step_count: int = Field(default=0, ge=0)
    recovery_count: int = Field(default=0, ge=0)
    intervention_count: int = Field(default=0, ge=0)
    pending_interrupt: dict[str, Any] | None = None
    terminal_reason: TerminalReason | None = None
    error_code: ErrorCode = ErrorCode.NONE


class RunResult(ClosedModel):
    schema_version: str = SCHEMA_VERSION
    run_id: UUID
    success: bool
    terminal_reason: TerminalReason
    error_code: ErrorCode = ErrorCode.NONE
    step_count: int = Field(ge=0)
    recovery_count: int = Field(ge=0)
    intervention_count: int = Field(ge=0)
    duration_ms: int = Field(ge=0)

File: apps/extension/src/contracts.ts
/** Stage 6 wire contracts mirrored from packages/agent/contracts.py. */

export const SCHEMA_VERSION = "1.0.0" as const;

export type ActionType =
  | "navigate"
  | "click"
  | "type"
  | "select"
  | "check"
  | "uncheck"
  | "press"
  | "scroll"
  | "wait"
  | "back"
  | "ask_user"
  | "handoff"
  | "stop";

export type VerificationStatus =
  | "UNVERIFIED"
  | "VERIFIED"
  | "FAILED"
  | "INCONCLUSIVE"
  | "UNCERTAIN"
  | "STALE";

export type RiskLevel = "LOW" | "MEDIUM" | "HIGH";
export type TerminalReason = "COMPLETED" | "USER_STOP" | "MAX_STEPS" | "SAFETY_STOP" | "ERROR";

export type ErrorCode =
  | "NONE"
  | "INVALID_ACTION"
  | "TARGET_NOT_FOUND"
  | "AMBIGUOUS_TARGET"
  | "STALE_OBSERVATION"
  | "TARGET_DISABLED"
  | "TARGET_NOT_VISIBLE"
  | "TARGET_NOT_EDITABLE"
  | "ACTION_TIMEOUT"
  | "NAVIGATION_INTERRUPTED"
  | "POLICY_BLOCKED"
  | "UNSUPPORTED_ACTION"
  | "EXECUTION_FAILED"
  | "VERIFICATION_FAILED"
  | "APPROVAL_REQUIRED"
  | "USER_TAKEOVER"
  | "MAX_STEPS_REACHED"
  | "INTERNAL_ERROR";

export interface UserCommand {
  schema_version: typeof SCHEMA_VERSION;
  command_id: string;
  session_id: string;
  text: string;
  modality: "text" | "voice_transcript";
  received_at: string;
}

export interface GoalSpec {
  schema_version: typeof SCHEMA_VERSION;
  task_id: string;
  objective: string;
  constraints: string[];
  allowed_actions: ActionType[];
  max_steps: number;
  risk_level: RiskLevel;
}

export interface AXNode {
  node_id: string;
  role: string;
  name: string;
  description: string | null;
  value_summary: string | null;
  states: Record<string, string | boolean | number>;
  level: number | null;
  disabled: boolean;
  focused: boolean;
  selected: boolean | null;
  checked: boolean | string | null;
  expanded: boolean | null;
  children: string[];
}

export interface Observation {
  schema_version: typeof SCHEMA_VERSION;
  observation_ref: string;
  version: number;
  url: string;
  title: string;
  captured_at: string;
  nodes: AXNode[];
  focused_node_id: string | null;
  content_hash: string;
  source: string;
  raw_char_count: number;
  compact_char_count: number;
  estimated_tokens: number;
  capture_latency_ms: number;
}

export interface AgentAction {
  schema_version: typeof SCHEMA_VERSION;
  step_id: string;
  action_type: ActionType;
  target_ref: string | null;
  input_value: string | null;
  key: string | null;
  observation_version: number;
  expected_effect: string;
  risk_level: RiskLevel;
  requires_approval: boolean;
}

export interface VerificationResult {
  schema_version: typeof SCHEMA_VERSION;
  verification_id: string;
  step_id: string;
  status: VerificationStatus;
  evidence: string[];
  before_observation_ref: string;
  after_observation_ref: string | null;
  checked_at: string;
  error_code: ErrorCode;
}

export interface RelevantItem {
  semantic_ref: string;
  label: string;
  reason: string;
  observation_version: number;
}

export interface TaskMapSnapshot {
  schema_version: typeof SCHEMA_VERSION;
  snapshot_id: string;
  session_id: string;
  run_id: string;
  version: number;
  observation_version: number;
  completed_items: string[];
  pending_items: string[];
  relevant_items: RelevantItem[];
  created_at: string;
}

export interface FocusHandoff {
  schema_version: typeof SCHEMA_VERSION;
  handoff_id: string;
  run_id: string;
  status: "NONE" | "REQUESTED" | "ACTIVE" | "RESUMING" | "COMPLETED";
  target_ref: string | null;
  announcement: string;
  created_at: string;
}

export interface AgentState {
  schema_version: typeof SCHEMA_VERSION;
  session_id: string;
  thread_id: string;
  run_id: string;
  task_id: string;
  goal: GoalSpec;
  constraints: string[];
  observation_version: number;
  task_map_version: number;
  active_semantic_ref: string | null;
  verification: VerificationResult | null;
  handoff_status: "NONE" | "REQUESTED" | "ACTIVE" | "RESUMING" | "COMPLETED";
  step_count: number;
  recovery_count: number;
  intervention_count: number;
  pending_interrupt: Record<string, unknown> | null;
  terminal_reason: TerminalReason | null;
  error_code: ErrorCode;
}

export interface RunResult {
  schema_version: typeof SCHEMA_VERSION;
  run_id: string;
  success: boolean;
  terminal_reason: TerminalReason;
  error_code: ErrorCode;
  step_count: number;
  recovery_count: number;
  intervention_count: number;
  duration_ms: number;
}

export type DisplayStatus = "VERIFIED_COMPLETED" | "PLANNED" | "UNCERTAIN" | "RELEVANT";

export interface TaskMapItem {
  item_id: string;
  label: string;
  status: DisplayStatus;
  semantic_ref: string | null;
  observation_version: number;
  verification_id: string | null;
  evidence: string[];
  reason: string | null;
}

export interface MapControlState {
  paused: boolean;
  takeover_active: boolean;
  approval_pending: boolean;
  handoff_status: "NONE" | "REQUESTED" | "ACTIVE" | "RESUMING" | "COMPLETED";
}

export interface AccessibleTaskMap {
  schema_version: "1.0.0";
  map_id: string;
  session_id: string;
  run_id: string;
  version: number;
  observation_version: number;
  goal: string;
  progress_label: string;
  verified_completed: TaskMapItem[];
  relevant_options: TaskMapItem[];
  next_action: TaskMapItem | null;
  uncertain_items: TaskMapItem[];
  control_state: MapControlState;
  final_summary: string | null;
  stale_invalidated_count: number;
  generated_at: string;
}

export type LiveRunStatus = "QUEUED" | "RUNNING" | "WAITING_USER" | "COMPLETED" | "FAILED" | "CANCELLED";

export interface LiveRunResponse {
  run_id: string;
  benchmark_session_id: string;
  task_id: string;
  status: LiveRunStatus;
  announcement: string;
  task_map: AccessibleTaskMap | null;
  error: string | null;
}

export interface ActiveBenchmarkTask {
  session_id: string;
  task_id: string;
  goal?: string;
  study_session_id?: string;
  instruction?: string;
  task_index?: number;
  task_count?: number;
}
